File: app/src/main/java/skarlat/dev/ecoproject/activity/CourseActivity.java
package skarlat.dev.ecoproject.activity;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.RecyclerView;

import java.io.IOException;
import java.util.List;

import skarlat.dev.ecoproject.ExtendedFunctionsKt;
import skarlat.dev.ecoproject.databinding.ActivityCourseCardBinding;
import skarlat.dev.ecoproject.includes.dataclass.Course;
import skarlat.dev.ecoproject.includes.dataclass.EcoCard;
import skarlat.dev.ecoproject.adapter.CardsViewAdapter;
import skarlat.dev.ecoproject.customView.ProgressBarView;
import skarlat.dev.ecoproject.includes.database.DatabaseHelper;

public class CourseActivity extends AppCompatActivity {
    private ProgressBarView progressBarView;
    private TextView leftCards;
    private TextView cursTitle;
    private TextView courseDesc;
    private RecyclerView recyclerView;
    private int progress;
    private Button startCourse;
    private String courseName;
    private DatabaseHelper db = new DatabaseHelper();
    private List<EcoCard> ecoCard;
    private Course currentCourse;
    private final int REQUST = 1;
    private ActivityCourseCardBinding binding;


    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityCourseCardBinding.inflate(getLayoutInflater());

        setContentView(binding.getRoot());

        Bundle tagView = getIntent().getExtras();

        currentCourse = db.getCourseByName(tagView.get("OPEN_COURSE").toString());
        courseName = currentCourse.getName();

        cursTitle = binding.cursTitle;
        progressBarView = binding.pbHorizontal;
        leftCards = binding.leftCards;
        courseDesc = binding.courseDesc;

        ecoCard = db.getAllCardsByCourseNameID(courseName);

        // получение ID по имени
        // int drawableID = this.getResources().getIdentifier(imgCourse, "drawable", getPackageName());
        // courseImgView.setBackgroundResource(drawableID);
        cursTitle.setText(currentCourse.getTitle());
        progress = currentCourse.getProgressBar();
        progress = db.getCourseByName(courseName).getProgressBar();
        progressBarView.setValue(progress);
        leftCards.setText(db.getLeftCards(courseName));
        courseDesc.setText(currentCourse.getFullDescription());

        ExtendedFunctionsKt.setImageFromAssets(binding.courseAvatar, getAssets(), currentCourse.pathBarImage());

//        if ( progress > 0  && progress < 100)
//            startCourse.setText("Продолжить обучение");
//        else if ( progress >= 100){
//            startCourse.setVisibility(View.GONE);
//        }
//        else
//            startCourse.setText("Начать обучение");


        CardsViewAdapter adapter = new CardsViewAdapter(CourseActivity.this, ecoCard, this::openCard);
        // TODO Add item decoration
        binding.recycleCards.setAdapter(adapter);
        db.upDateIsCurrentCourse(courseName);
    }


    @Override
    public void startActivityForResult(Intent intent, int requestCode) {
        super.startActivityForResult(intent, requestCode);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        // TODO("LOGIC FOR CLOSE CARD")
//        progress = currentCourse.getProgressBar();
//        progressBarView.setValue(progress);
//        leftCards.setText(db.getLeftCards(courseName));
//
//        CardsViewAdapter adapter = new CardsViewAdapter(CourseCardActivity.this, ecoCard);
//        recyclerView.setAdapter(adapter);
//
//        if ( progress > 0  && progress < 100)
//            startCourse.setText("Продолжить обучение");
//        else if ( progress >= 100){
//            startCourse.setVisibility(View.GONE);
//        }
//        else
//            startCourse.setText("Начать обучение");
        super.onActivityResult(requestCode, resultCode, data);
    }

    public void onBackBtn(View view) {
        onBackPressed();
    }


    @Override
    public void onBackPressed() {
        super.onBackPressed();
    }

    /**
     * View карточки обновляет базу данных при переходе
     *
     * @param view
     */
    public Void openCard(View view) {
        EcoCard currentCard = (EcoCard) view.getTag();

        Intent intent = new Intent(this, CardActivity.class);
        intent.putExtra(currentCard.getClass().getSimpleName(), currentCard);
        if (currentCard.getStatus() == EcoCard.Status.WATCHED) {
            startActivityForResult(intent, REQUST);
        } else {

            for (int i = 0; i < ecoCard.size(); i++) {
                if (ecoCard.get(i).getName() == currentCard.getName() && (i + 1) < ecoCard.size()) {
                    ecoCard.get(i + 1).upDate(EcoCard.Status.OPENED);
                    db.updateFirebaseProgress("Cards", ecoCard.get(i + 1).getName(), "status", 1);
                    break;
                }
            }

            currentCard.upDate(EcoCard.Status.WATCHED);
            db.updateFirebaseProgress("Cards", currentCard.getName(), "status", 2);
            upDateCurrentCourse();

            startActivityForResult(intent, REQUST);
        }
        return null;
    }

    /**
     * Progress бар для базы данных
     */
    private void upDateCurrentCourse() {

        double res = 100.00 / (double) ecoCard.size();
        res = Math.ceil(res);

        progress += res;
        if (progress >= 100) {
            currentCourse.upDate(progress, Course.Status.FINISHED);
        } else {
            currentCourse.upDate(progress, Course.Status.CURRENT);
        }
        db.updateFirebaseProgress("Courses", currentCourse.getName(), "progress", progress);
    }

    /**
     * Кнопка продолжить обучение
     * Обновляет базу данных при нажатии
     *
     * @param view
     */
    public void startBtn(View view) {
        for (int i = 0; i < ecoCard.size(); i++) {
            EcoCard card = ecoCard.get(i);
            Enum status = card.getStatus();
            if (status == EcoCard.Status.OPENED) {
                Intent intent = new Intent(this, CardActivity.class);

                intent.putExtra(card.getClass().getSimpleName(), card);

                upDateCurrentCourse();

                if (i != ecoCard.size() - 1) {
                    ecoCard.get(i + 1).upDate(EcoCard.Status.OPENED);
                    db.updateFirebaseProgress("Cards", ecoCard.get(i + 1).getName(), "status", 1);
                }
                card.upDate(EcoCard.Status.WATCHED);
                db.updateFirebaseProgress("Cards", card.getName(), "status", 2);
                startActivityForResult(intent, REQUST);
                break;
            }
        }
    }
}

File: app/src/main/java/skarlat/dev/ecoproject/activity/CourseActivity.kt
package skarlat.dev.ecoproject.activity

import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import by.kirich1409.viewbindingdelegate.viewBinding
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import skarlat.dev.ecoproject.*
import skarlat.dev.ecoproject.adapter.CardsViewAdapter
import skarlat.dev.ecoproject.databinding.ActivityCourseCardBinding
import skarlat.dev.ecoproject.includes.database.DatabaseHelper
import skarlat.dev.ecoproject.includes.dataclass.Course
import skarlat.dev.ecoproject.includes.dataclass.Course.Status.FINISHED
import skarlat.dev.ecoproject.includes.dataclass.EcoCard

class CourseActivity : AppCompatActivity() {
    private var courseName: String? = null
    private val db = DatabaseHelper()
    private var ecoCards: List<EcoCard>? = null
    private var currentCourse: Course? = null
    private val REQUST = 1
    private val binding: ActivityCourseCardBinding by viewBinding()
    private var adapter: CardsViewAdapter? = null

    // TODO Refactor it
    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(binding.root)
        adapter = CardsViewAdapter { view: View -> openCard(view) }
        updateData()
        binding!!.recycleCards.adapter = adapter
        binding!!.courseAvatar.setImageFromAssets(assets, currentCourse!!.pathBarImage())
        binding!!.recycleCards.addItemDecoration(ConnectionCardDecorator())
        if (currentCourse?.isActive == Course.Status.CLOSED.ordinal) {
            markCourseAsCurrent()
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == Const.CARD_OPENED) {
            if (resultCode == Const.CARD_ACTIVITY_OK) {
                updateData()
            }
        }
        super.onActivityResult(requestCode, resultCode, data)
    }

    fun onBackBtn(view: View?) {
        onBackPressed()
    }

    /**
     * View карточки обновляет базу данных при переходе
     *
     * @param view
     */
    fun openCard(view: View): Void? {
        val currentCard = view.tag as EcoCard
        val intent = Intent(this, CardActivity::class.java)
        intent.putExtra(currentCard.javaClass.simpleName, currentCard)
        val nextCard = ecoCards?.itemAfter(currentCard)

        if (currentCard.status === EcoCard.Status.WATCHED && nextCard?.status != EcoCard.Status.CLOSED) {
            startActivityForResult(intent, REQUST)
        } else {
            currentCard.setStatus(EcoCard.Status.WATCHED)
            nextCard?.setStatus(EcoCard.Status.OPENED)
            updateCard(currentCard)
            startActivityForResult(intent, REQUST)
        }
        if (nextCard == null) {
            markCourseAsFinished()
        } else {
            updateCard(nextCard)
        }
        return null
    }

    private fun updateCard(card: EcoCard?) {
        lifecycleScope.launch(Dispatchers.IO) {
            EcoTipsApp.getDatabase().cardsDao().update(card)
        }
    }

    private fun markCourseAsFinished() {
        currentCourse?.let {
            lifecycleScope.launch {
                it.isActive = FINISHED.ordinal
                EcoTipsApp.getDatabase().courseDao().update(it)
            }
        }
    }

    private fun markCourseAsCurrent() {
        currentCourse?.let {
            lifecycleScope.launch {
                it.isActive = Course.Status.CURRENT.ordinal
                EcoTipsApp.getDatabase().courseDao().update(it)
            }
        }
    }

    override fun onResume() {
        super.onResume()
        updateData()
    }

    private fun updateData() {
        val tagView = intent.extras
        currentCourse = db.getCourseByName(tagView!!["OPEN_COURSE"].toString())
        courseName = currentCourse?.getName()
        ecoCards = db.getAllCardsByCourseNameID(courseName)
        currentCourse?.leftCards = ecoCards?.count { it.status != EcoCard.Status.WATCHED } ?: 0
        currentCourse?.countCards = ecoCards?.size ?: 0
        val adapterList = arrayListOf<Any>(currentCourse!!)
        adapterList.addAll(ecoCards!!)
        adapter!!.submitList(adapterList)
    }
}